### ISSUE TITLE
Cannot find a provider for deploy_revision[nodejs application] on ubuntu version 12.04

### ISSUE THREAD

**Issue by marcoamorales** [rackspace-cookbooks/nodestack#173]
In Chef 12.0 I get the following error message when trying to converge.

I'm pretty sure this is related to the change in Chef 12: `Added ProviderResolver feature to handle late resolution of providers based on what kinds of support is in the base operating system.`

```
================================================================================       
  Error executing action `deploy` on resource 'deploy_revision[nodejs application]'       
  ================================================================================       
         
  ArgumentError       
  -------------       
  Cannot find a provider for deploy_revision[nodejs application] on ubuntu version 12.04       
         
  Cookbook Trace:       
  ---------------       
  /tmp/kitchen/cache/cookbooks/application/libraries/default.rb:144:in `deploy_provider'       
  /tmp/kitchen/cache/cookbooks/application/libraries/default.rb:152:in `release_path'       
  /tmp/kitchen/cache/cookbooks/nodestack/recipes/application_nodejs.rb:92:in `block (3 levels) in from_file'       
         /tmp/kitchen/cache/cookbooks/application/libraries/default.rb:189:in `safe_recipe_eval'
         /tmp/kitchen/cache/cookbooks/application/libraries/default.rb:165:in `callback'
         /tmp/kitchen/cache/cookbooks/application/providers/default.rb:182:in `run_actions_with_context'
         /tmp/kitchen/cache/cookbooks/application/providers/default.rb:161:in `block (2 levels) in run_deploy'
         
       
         Resource Declaration:
         ---------------------
       
         # In /tmp/kitchen/cache/cookbooks/application/providers/default.rb
         
         123:   @deploy_resource = send(new_resource.strategy.to_sym, new_resource.name) do
         124:     action force ? :force_deploy : :deploy
       
         125:     scm_provider new_resource.scm_provider
         126:     revision new_resource.revision
         127:     repository new_resource.repository
         128:     enable_submodules new_resource.enable_submodules
         129:     user new_resource.owner
         130:     group new_resource.group
       
         131:     deploy_to new_resource.path
         132:     ssh_wrapper "#{new_resource.path}/deploy-ssh-wrapper" if new_resource.deploy_key
         133:     shallow_clone new_resource.shallow_clone
         134:     rollback_on_error new_resource.rollback_on_error
         135:     all_environments = ([new_resource.environment]+new_resource.sub_resources.map{|res| res.environment}).inject({}){|acc, val| acc.merge(val)}
       
         136:     environment all_environments
         137:     migrate new_resource.migrate
         138:     all_migration_commands = ([new_resource.migration_command]+new_resource.sub_resources.map{|res| res.migration_command}).select{|cmd| cmd && !cmd.empty?}
         139:     migration_command all_migration_commands.join(' && ')
         140:     restart_command do
         141:       ([new_resource]+new_resource.sub_resources).each do |res|
         142:         cmd = res.restart_command
       
         143:         if cmd.is_a? Proc
         144:           app_provider.deploy_provider.instance_eval(&cmd) # @see libraries/default.rb
         145:         elsif cmd && !cmd.empty?
         146:           execute cmd do
         147:             user new_resource.owner
         148:             group new_resource.group
       
         149:             environment all_environments
         150:           end
         151:         end
         152:       end
         153:     end
         154:     purge_before_symlink (new_resource.purge_before_symlink + new_resource.sub_resources.map(&:purge_before_symlink)).flatten
         
       
         Compiled Resource:
         ------------------
         # Declared in /tmp/kitchen/cache/cookbooks/application/providers/default.rb:123:in `run_deploy'
         
         deploy_revision("nodejs application") do
           action [:deploy]
           updated true
           updated_by_last_action true
           retries 0
           retry_delay 2
           default_guard_interpreter :default
           deploy_to "/var/app"
           repository_cache "cached-copy"
       
    revision "HEAD"       
           rollback_on_error true
           remote "origin"
       
           shallow_clone true
           scm_provider Chef::Provider::Git
         
         keep_releases 5
         
         enable_checkout true
           checkout_branch "deploy"
           cookbook_name "nodestack"
           repo "git@github.com:marcoamorales/node-hello-world.git"
           user "my_nodejs_app"
       
           group "my_nodejs_app"
           git_ssh_wrapper "/var/app/deploy-ssh-wrapper"
           restart_command #<Proc:0x00000004e49bf8@/tmp/kitchen/cache/cookbooks/application/providers/default.rb:140>
           before_migrate #<Proc:0x00000004e49860@/tmp/kitchen/cache/cookbooks/application/providers/default.rb:160>
           before_symlink #<Proc:0x00000004e4b4f8@/tmp/kitchen/cache/cookbooks/application/providers/default.rb:163>
       
           before_restart #<Proc:0x00000004e51128@/tmp/kitchen/cache/cookbooks/application/providers/default.rb:166>
           after_restart #<Proc:0x00000004e53b58@/tmp/kitchen/cache/cookbooks/application/providers/default.rb:169>
           shared_path "/var/app/shared"
           destination "/var/app/shared/cached-copy"
           current_path "/var/app/current"
       
         end
         
       
[2014-12-08T20:40:00+00:00] INFO: Running queued delayed notifications before re-raising exception       
       
Running handlers:       
[2014-12-08T20:40:00+00:00] ERROR: Running exception handlers       
Running handlers complete       
[2014-12-08T20:40:00+00:00] ERROR: Exception handlers complete       
[2014-12-08T20:40:00+00:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out       
Chef Client failed. 40 resources updated in 235.514875959 seconds       
[2014-12-08T20:40:00+00:00] ERROR: deploy_revision[nodejs application] (/tmp/kitchen/cache/cookbooks/application/providers/default.rb line 123) had an error: ArgumentError: Cannot find a provider for deploy_revision[nodejs application] on ubuntu version 12.04       
[2014-12-08T20:40:01+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1) 
```

**Comment by martinb3** [comment]
Tagging with https://github.com/AutomationSupport/stack-issues/issues/8.

**Comment by marcoamorales** [comment]
Fixed in #183
